hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/bulk/RowDataKeyGen.java

```diff
@@ -99,7 +99,7 @@ protected RowDataKeyGen(
       this.recordKeyProjection = null;
     } else {
       this.recordKeyFields = recordKeys.get().split(",");
-      if (this.recordKeyFields.length == 1) {
+      if (this.recordKeyFields.length == 1  && this.partitionPathFields.length == 1) {
```

Comment:
Nice findings ~

I kind of think the handling of `RowDataKeyGen` is correct, what we need to fix is the `ComplexAvroKeyGenerator`, it composite the record key string as `key_name:key_val` even if the record key has only one field.

If we take a look at the `ComplexKeyGenerator.getRecordKey(InternalRow internalRow, StructType schema)`, it also eliminate the `key_name` from the record key string even if there are multiple partiton fields. And for compatible interface `getRecordKey(GenericRecord record)`, it uses the delegate generator `ComplexAvroKeyGenerator`, which causes the inconsistency of the two interfaces.

In summary, I think `key_name:key_val` should be a mistake when the record key fields is single, it should be simplified as `key_val`.

Reply:
Got it

Reply:
Make  SimpleAvroKeyGenerator  compatible with multiple partition. ?
Or 
Make ComplexAvroKeyGenerator compatible with single partition?

Reply:
Nice catch here @hehuiyuan @danny0405 
We also also need to consider historical data compatibility issues. For example users already have hudi table using `key_id: key` as hoodie_key

Reply:
There is another compatibility issue with existing table.    The _hoodie_record_key  of same key is changed when fix the issure by modifying upsert mode.

Reply:
> continue to use ComplexAvroKeyGenerator make ComplexAvroKeyGenerator compatible with single primary key

I'm approval with this way for 2 reasons:

1. The ComplexKeyGenerator for row writers already did this;
2. There is no need to change the strategy of what key generator to use in general, to avoid many potential changes.

Reply:
> We also also need to consider historical data compatibility issues. For example users already have hudi table using key_id: key as hoodie_key

@danny0405 @zhangyue19921010
Indeed this PR likely has introduced duplicates in our tables when migrating from 0.13 to 0.14 see #10508